core/service-test-runner/icedfrisby-shields.js

```diff
@@ -32,18 +33,21 @@ const factory = superclass =>
     }
 
     expectBadge({ label, message, color }) {
-      let expectedBadge
-      if (typeof color === 'undefined') {
-        expectedBadge = { name: label, value: message }
-      } else {
-        expectedBadge = { name: label, value: message, color }
-      }
-
-      if (typeof message === 'string') {
-        return this.expectJSON(expectedBadge)
-      } else {
-        return this.expectJSONTypes(Joi.object().keys(expectedBadge))
-      }
+      return this.afterJSON(json => {
+        expect(json.label, 'label mismatch').to.equal(label)
```

Comment:
One option to consider here is using multiple Frisby expectations instead of `afterJSON()`. The advantage of Frisby `expect`s is that they all run, so if multiple things go wrong, they are all reported. If they’re all chained into an `after` the test stops after the first failure.

I haven’t used that part of the API much myself so I can’t vouch for its ergonomics, though we can certainly work to improve it if it turns out there are problems.

In particular, I think you can call [`expectJSON()`](https://github.com/IcedFrisby/IcedFrisby/blob/master/API.md) (or `expectJSONTypes()`) multiple times, using the optional `path` argument:

```js
this.expectJSONTypes(‘label’, label)
this.expectJSONTypes(‘message’, message)
if (typeof color !== undefined) {
  this.expectJSONTypes(‘color’, color)
}
return this
```

I’m certainly fine with what you’ve done here as well! This could definitely be tackled / considered in a follow-up.

Reply:
Using `path` was my initial attempt before the [issue with afterJSON](https://github.com/IcedFrisby/IcedFrisby/issues/206) was fixed, but if I remember correctly it didn't work as the second argument (`jsonTest`) must be an object rather than a plain string.

I also tried bundling everything into a single expect, but that didn't work nicely either:
* `expectJSON` does strict matching, so we would have to specify additional fields such as `color` for every single test. Services use a lot of different colors for the messages, so there's no sensible default value we could set. In my opinion, adding colors everywhere would add noise to our test suite rather than improve it.
* `expectContainsJSON` is a looser alternative, but when the match fails the error message isn't as clear as "expected message to be foo but was bar". In particular, it will highlight the deprecated fields `value` and `name` which we don't care about. Even though I did get the assertions working, the developer experience was quite confusing.

Reply:
Make sense. Want to open an issue documenting some of this?

Reply:
Will do so tomorrow.